src/ucp/tag/rndv.c

```diff
@@ -283,21 +385,165 @@ static ucs_status_t
+    } else {
+        /* send multiple bcopy messages (fragments of the original send message) */
+        ucs_trace_req("Progress send request %p on am lane: %d. multi message (bcopy)",
+                      sreq, sreq->send.lane);
+        status = ucp_do_am_bcopy_multi(self, UCP_AM_ID_RNDV_DATA,
+                                       UCP_AM_ID_RNDV_DATA,
+                                       UCP_AM_ID_RNDV_DATA_LAST,
+                                       sizeof(ucp_rndv_data_hdr_t),
+                                       ucp_rndv_pack_multi_data,
+                                       ucp_rndv_pack_multi_data,
+                                       ucp_rndv_pack_multi_data_last);
+    }
+    if (status == UCS_OK) {
+        ucp_request_complete_send(sreq, UCS_OK);
+    }
+
+    return status;
+}
+
+static ucs_status_t
+ucp_rndv_rtr_handler(void *arg, void *data, size_t length, void *desc)
+{
+    ucp_rndv_rtr_hdr_t *rndv_rtr_hdr = data;
+    ucp_request_t *sreq = (ucp_request_t*) rndv_rtr_hdr->sreq_ptr;
+
+    /* make sure that the ep on which the rtr was received on is connected */
+    ucs_assert_always(!ucp_ep_is_stub(sreq->send.ep));
+    ucs_trace_req("received RTR request. Start sending on send request %p", sreq);
+
+    /* dereg the original send request since we are going to send with bcopy */
+    ucp_request_send_buffer_dereg(sreq, ucp_ep_get_rndv_data_lane(sreq->send.ep));
+
+    sreq->send.uct.func   = ucp_rndv_progress_send;
+    sreq->send.proto.rreq = rndv_rtr_hdr->rreq_ptr;
+
+    ucp_request_start_send(sreq);
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_rndv_data_handler(void *arg, void *data, size_t length, void *desc)
+{
+    ucp_rndv_data_hdr_t *rndv_data_hdr = data;
+    ucp_request_t *rreq = (ucp_request_t*) rndv_data_hdr->rreq_ptr;
+    size_t recv_len;
+    ucs_status_t status;
+    uint8_t hdr_len = sizeof(*rndv_data_hdr);
+
+    ucs_trace_req("received segment for request %p. offset %zu ",
+                   rreq, rreq->recv.state.offset);
+
+    ucs_assert(length >= hdr_len);
+    recv_len = length - hdr_len;
+    status = ucp_tag_process_recv(rreq->recv.buffer, rreq->recv.count,
+                                  rreq->recv.datatype, &rreq->recv.state,
+                                  data + hdr_len, recv_len, 0);
+    rreq->recv.state.offset += recv_len;
+
+    return status;
+}
+
+static ucs_status_t
+ucp_rndv_data_last_handler(void *arg, void *data, size_t length, void *desc)
+{
+    ucp_rndv_data_hdr_t *rndv_data_hdr = data;
+    ucp_request_t *rreq = (ucp_request_t*) rndv_data_hdr->rreq_ptr;
+    size_t recv_len;
+    ucs_status_t status;
+    uint8_t hdr_len = sizeof(*rndv_data_hdr);
+
+    ucs_trace_req("received last segment for request %p offset %zu with ",
+                   rreq, rreq->recv.state.offset);
+
+    ucs_assert(length >= hdr_len);
+    recv_len = length - hdr_len;
+    status = ucp_tag_process_recv(rreq->recv.buffer, rreq->recv.count,
```

Comment:
since we can't tell on which am_id the handler was called, i can't unite the functions. 
extracting common code in this case would mean to call many of the parameters here in a third function. also, the usage of recv_len is different between the functions. 
do you think that it would be better with a third function in this case?

rest of the CR comments were applied.